File: 01-scenarios/Service-to-Service-Agent/0.Resources/Starting-point/nodejs/src/tools.ts
import type { InventoryDetails, OrderDetails, RunContext, ShipmentEvent } from "./types.js";

export interface IOrderReader {
    read(orderId: string, run: RunContext, signal: AbortSignal): Promise<OrderDetails>;
}
export interface IInventoryReader {
    read(sku: string, run: RunContext, signal: AbortSignal): Promise<InventoryDetails>;
}
export interface IChannelNotifier {
    notify(shipment: ShipmentEvent, summary: string, run: RunContext, signal: AbortSignal): Promise<string>;
}
export interface ITicketWriter {
    create(shipment: ShipmentEvent, summary: string, run: RunContext, signal: AbortSignal): Promise<string>;
}

export class StubOrderReader implements IOrderReader {
    async read(orderId: string, _run: RunContext, signal: AbortSignal): Promise<OrderDetails> {
        signal.throwIfAborted();
        return { orderId, sku: "WIDGET-42", quantity: 20, destination: "Milan" };
    }
}

export class StubInventoryReader implements IInventoryReader {
    async read(sku: string, _run: RunContext, signal: AbortSignal): Promise<InventoryDetails> {
        signal.throwIfAborted();
        return { sku, availableUnits: 50, warehouse: "Bergamo" };
    }
}

export class StubChannelNotifier implements IChannelNotifier {
    async notify(_shipment: ShipmentEvent, _summary: string, run: RunContext, signal: AbortSignal): Promise<string> {
        signal.throwIfAborted();
        return `stub-notification-${run.runId}`;
    }
}

export class StubTicketWriter implements ITicketWriter {
    async create(_shipment: ShipmentEvent, _summary: string, run: RunContext, signal: AbortSignal): Promise<string> {
        signal.throwIfAborted();
        return `stub-ticket-${run.runId}`;
    }
}

File: 01-scenarios/Service-to-Service-Agent/0.Resources/Starting-point/python/app/tools.py
from dataclasses import dataclass
from typing import Protocol

from app.models import RunContext, ShipmentEvent


@dataclass(frozen=True)
class OrderDetails:
    order_id: str
    sku: str
    quantity: int
    destination: str


@dataclass(frozen=True)
class InventoryDetails:
    sku: str
    available_units: int
    warehouse: str


class OrderReader(Protocol):
    async def read(self, order_id: str, run: RunContext) -> OrderDetails: ...


class InventoryReader(Protocol):
    async def read(self, sku: str, run: RunContext) -> InventoryDetails: ...


class ChannelNotifier(Protocol):
    async def notify(self, shipment: ShipmentEvent, summary: str, run: RunContext) -> str: ...


class TicketWriter(Protocol):
    async def create(self, shipment: ShipmentEvent, summary: str, run: RunContext) -> str: ...


class StubOrderReader:
    async def read(self, order_id: str, run: RunContext) -> OrderDetails:
        return OrderDetails(order_id, "WIDGET-42", 20, "Milan")


class StubInventoryReader:
    async def read(self, sku: str, run: RunContext) -> InventoryDetails:
        return InventoryDetails(sku, 50, "Bergamo")


class StubChannelNotifier:
    async def notify(self, shipment: ShipmentEvent, summary: str, run: RunContext) -> str:
        return f"stub-notification-{run.run_id}"


class StubTicketWriter:
    async def create(self, shipment: ShipmentEvent, summary: str, run: RunContext) -> str:
        return f"stub-ticket-{run.run_id}"
